### PR TITLE
feat(op-service):Persist RethDB instance in the go fetcher struct.

### PR DESCRIPTION
**Pull request by Nickqiaoo**
solve for https://github.com/ethereum-optimism/optimism/pull/9349

### CODE REVIEW

**Comment by Nickqiaoo** [comment]
@clabby @sebastianst Please review🙂

**Review by clabby — APPROVED** [review]
Rust changes LGTM. Deferring to @sebastianst for final approval of the Go changes. Thank you @Nickqiaoo 🥳

**Review by trianglesphere — APPROVED** [review]

**Comment by Nickqiaoo** [comment]
Could you please review the code when you have free time🥹 @sebastianst

File: op-service/sources/reth_db.go [hunk 1]
```diff
@@ -68,26 +71,40 @@ func FetchRethReceipts(dbPath string, blockHash *common.Hash) (types.Receipts, e
...
+func OpenDBReadOnly(dbPath string) (db unsafe.Pointer, err error) {
+	// Convert the db path to a C string and defer its deallocation
+	cDbPath := C.CString(dbPath)
+	defer C.free(unsafe.Pointer(cDbPath))
+
+	// Call the C function to fetch the receipts from the Reth Database
+	openDBResult := C.open_db_read_only(cDbPath)
+
+	if openDBResult.error {
+		return nil, fmt.Errorf("Error open db from Reth Database.")
```
Comment (trianglesphere) [inline op-service/sources/reth_db.go]:
```suggestion
		return nil, fmt.Errorf("failed to open RethDB")
```

File: op-service/sources/reth_db.go [hunk 2]
```diff
@@ -68,26 +71,40 @@ func FetchRethReceipts(dbPath string, blockHash *common.Hash) (types.Receipts, e
...
-	//   which would be more optimal.
-	//   We could move the opening of the RethDB and creation of the db reference
-	//   into NewRethDBReceiptsFetcher.
+	dbInstance unsafe.Pointer
 }
 
 var _ ReceiptsProvider = (*RethDBReceiptsFetcher)(nil)
 
 func NewRethDBReceiptsFetcher(dbPath string) *RethDBReceiptsFetcher {
+	db, _ := OpenDBReadOnly(dbPath)
```
Comment (trianglesphere) [inline op-service/sources/reth_db.go]: Could we make this function faillible (i.e. it returns `(*RethDBReceiptsFetcher, error) ` and report the error earlier?

Comment (Nickqiaoo) [reply op-service/sources/reth_db.go]: if we do this, we need also make `newRecProviderFromConfig` return error, and check error in `NewEthClient` in `eth_client.go` , and always return error with nil in `receipts_rpc.go`.It seems feasible

> Could we make this function faillible (i.e. it returns `(*RethDBReceiptsFetcher, error) ` and report the error earlier?

Comment (Nickqiaoo) [reply op-service/sources/reth_db.go]: @trianglesphere  I use a more simple way to implement this. Return nil when error and check nil in the `NewEthClient` function

> Could we make this function faillible (i.e. it returns `(*RethDBReceiptsFetcher, error) ` and report the error earlier?

File: op-service/sources/reth_db.go [hunk 3]
```diff
@@ -93,15 +93,18 @@ type RethDBReceiptsFetcher struct {
 var _ ReceiptsProvider = (*RethDBReceiptsFetcher)(nil)
 
 func NewRethDBReceiptsFetcher(dbPath string) *RethDBReceiptsFetcher {
```
Comment (trianglesphere) [inline op-service/sources/reth_db.go:96]:
```suggestion
// NewRethDBReceiptsFetcher opens a RethDB for reading receipts. It returns nil if it was unable to open the database
func NewRethDBReceiptsFetcher(dbPath string) *RethDBReceiptsFetcher {
```

**Review by trianglesphere — APPROVED** [review]

**Pull request merged**